base/set.jl

```diff
@@ -123,16 +123,46 @@ julia> unique([1; 2; 2; 6])
  6
 ```
 """
-function unique(C)
-    out = Vector{eltype(C)}()
-    seen = Set{eltype(C)}()
-    for x in C
+function unique(itr)
+    T = _default_eltype(typeof(itr))
+    out = Vector{T}()
+    seen = Set{T}()
+    i = start(itr)
+    if done(itr, i)
+        return out
+    end
+    x, i = next(itr, i)
+    if !isleaftype(T)
+        T = typeof(x)
+        return unique_from(itr, Vector{T}(), Set{T}(), x, i)
+    end
+    push!(seen, x)
+    push!(out, x)
+    if !done(itr, i)
+        x, i = next(itr, i)
+    end
+    return _unique(itr, out, seen, x, i)
```

Comment:
That was my thinking, but somehow I got the idea from a question I asked @yuyichao that the body of `unique_from` would also be inlined. Maybe I assumed it would be the case and it is not.